jws/jws.go

```diff
@@ -424,40 +431,64 @@ func ParseString(s string) (*Message, error) {
 	return Parse(strings.NewReader(s))
 }
 
-func parseJSON(src io.Reader) (result *Message, err error) {
+type fullMessageProxy struct {
+	// encoded signature fields
+	Signature json.RawMessage `json:"signature"`
+	Headers   json.RawMessage `json:"header"` // あれ、"s"いらないんだっけ
+	Protected json.RawMessage `json:"protected"`
 
-	var wrapper FullEncodedMessageUnmarshalProxy
+	// encoded message fields
+	Signatures []*encodedSignature `json:"signatures"`
+	Payload    string              `json:"payload"`
+}
 
-	if err := json.NewDecoder(src).Decode(&wrapper); err != nil {
-		return nil, errors.Wrap(err, `failed to unmarshal jws message`)
+func (proxy *fullMessageProxy) encodedSignature() (*encodedSignature, error) {
+	var encodedSig encodedSignature
+	if err := json.Unmarshal(proxy.Protected, &encodedSig.Protected); err != nil {
+		return nil, errors.Wrap(err, `failed to unmarshal 'protected' field`)
 	}
+	if err := json.Unmarshal(proxy.Signature, &encodedSig.Signature); err != nil {
+		return nil, errors.Wrap(err, `failed to unmarshal 'signature' field`)
+	}
+	h := NewHeaders()
+	if err := json.Unmarshal(proxy.Headers, h); err != nil {
+		return nil, errors.Wrap(err, `failed to unmarshal 'header' field`)
+	}
+
+	return &encodedSig, nil
+}
 
-	if wrapper.EncodedMessageUnmarshalProxy == nil {
-		return nil, errors.New(`invalid payload (probably empty)`)
+func parseJSON(src io.Reader) (result *Message, err error) {
+	var proxy fullMessageProxy
+	if err := json.NewDecoder(src).Decode(&proxy); err != nil {
+		return nil, errors.Wrap(err, `failed to unmarshal jws message`)
 	}
 
-	// if the "signature" field exist, treat it as a flattened
-	if wrapper.EncodedSignatureUnmarshalProxy != nil {
-		if len(wrapper.Signatures) != 0 {
+	if len(proxy.Signature) > 0 {
+		if len(proxy.Signatures) > 0 {
 			return nil, errors.New("invalid message: mixed flattened/full json serialization")
```

Comment:
With "flattened" you mean the "dot"-style format? According to the RFC (https://tools.ietf.org/html/rfc7515) the terminology is "compact serialization" (with the dots) and "JSON serialization"